src/include/OpenImageIO/refcnt.h

```diff
@@ -105,6 +105,25 @@ class intrusive_ptr
         }
     }
 
+    /// Set this smart pointer to null, decrement the object's reference
+    /// count, return the original raw pointer, but do NOT delete the object
+    /// even if the ref count goes to zero. The only safe use case is to
+    /// convert the sole managed pointer to an object into a raw pointer.
+    /// DANGER -- use with caution! This is only safe to do if no other
+    /// intrusive_ptr refers to the object (such a pointer may subsequently
+    /// reset, decrementing the count to 0, and incorrectly free the
+    /// object), and it can cause a memory leak if the caller isn't careful
+    /// to either reassign the returned pointer to another managed pointer
+    /// or delete it manually.
+    T* release () {
```

Comment:
This is mostly intended to be used in cases where we are sure that we are the "last" owner right?

Maybe we could assert the ref count is 1 when we start?

Reply:
Yes, the only safe case is when we are the only owner.
I can assert, but it's still not 100% thread-safe (if it's being misused) because between the time the refcnt is accessed and when it's checked, another thread could add a reference.

Reply:
Added assertion and pushed update.

Reply:
You are right. It seems like a `DASSERT` would at least help catch the "normal" uses of this (like in the OSL example) while occasionaly firing if you were using this in some situation you probably shouldn't.

Maybe checking for 0 after the decref is even more reliable? (since no two threads should ever "see" 0 at the same time)

Reply:
I think I would suggest the following:

```
if (!p->_decref ())
   DASSERT_MSG(false, "trying to use release without being the last owner");
```

Reply:
I think that's equivalent, but I kind of like that way of using the result of _decref.

Reply:
Updated again

Reply:
LGTM!